### DISCUSSION TITLE
Convert markdown to mdast tree with gfm extention don't work (Cannot set properties of undefined (setting 'inTable'))

### DISCUSSION THREAD

**Discussion by nicolasjorgevera** [syntax-tree/unist#117] [Q&A] [answered]
I try to convert an .md file to a mdast tree(JSON) using a gfm extention in Next.js 13 and I got an Typescript in VSC and Next.js "Cannot set properties of undefined (setting 'inTable')" error in the browser.

The code is the next one:

```typescript
"use server"
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import { fromMarkdown } from "mdast-util-from-markdown"
import { gfm } from 'micromark-extension-gfm'
import { gfmFromMarkdown, gfmToMarkdown } from 'mdast-util-gfm'

const postsDirectory = path.join(process.cwd(), 'blogposts');

export async function getPostDataToEdit (id: string) {
  const fullPath = path.join(postsDirectory, `${id}.md`);
  const fileContents = fs.readFileSync(fullPath, 'utf8');
  const matterResult = matter(fileContents);

  const tree = fromMarkdown(fileContents,'utf8' , {
    extensions: [gfm()],
    mdastExtensions: [ gfmFromMarkdown() ]
  })

  const blogPost: BlogPost & {  body: any } = {
    id,
    title: matterResult.data.title,
    date: matterResult.data.date,
    published: matterResult.data.published ? "Published" : "Draft",
    body: tree,
  }

  return blogPost;
}
``` 

The typescript error is:
![typescript error](https://i.stack.imgur.com/gIQ0g.jpg)

The Next.js error is:
![next.js error](https://i.stack.imgur.com/cI7Do.jpg)

I read the documentation of mdast-util-from-markdown and mdast-util-gfm and couldn't solve this. If I remove the extention section from the function "fromMarkdown" works just fine but without idenifying the gfm markdown on the file, just only common markdown.

The code without extention section:

```typescript
"use server"
import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';
import { fromMarkdown } from "mdast-util-from-markdown"
import { gfm } from 'micromark-extension-gfm'
import { gfmFromMarkdown, gfmToMarkdown } from 'mdast-util-gfm'

const postsDirectory = path.join(process.cwd(), 'blogposts');

export async function getPostDataToEdit (id: string) {
  const fullPath = path.join(postsDirectory, `${id}.md`);
  const fileContents = fs.readFileSync(fullPath, 'utf8');
  const matterResult = matter(fileContents);

  const tree = fromMarkdown(fileContents,'utf8' , {
  })

  const blogPost: BlogPost & {  body: any } = {
    id,
    title: matterResult.data.title,
    date: matterResult.data.date,
    published: matterResult.data.published ? "Published" : "Draft",
    body: tree,
  }

  return blogPost;
}
``` 

Please :pray:, can you help me? 

Thanks!!

**Comment by wooorm** [comment]
can you post versions?

Can you post `fileContents`? Important: first make `fileContents` as small as possible to reproduce the error.

**Reply by nicolasjorgevera** [reply]
Hello wooorm, I send you what you requested.

```typescript
{
  "name": "blog-base-app",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@mdx-js/loader": "^2.3.0",
    "@mdx-js/react": "^2.3.0",
    "@next/mdx": "^13.4.3",
    "@nextui-org/react": "^1.0.0-beta.12",
    "@radix-ui/react-dropdown-menu": "^2.0.5",
    "@radix-ui/react-slot": "^1.0.2",
    "@reduxjs/toolkit": "^1.9.5",
    "@tanstack/react-table": "^8.9.3",
    "@types/mdast": "^4.0.0",
    "@types/mdx": "^2.0.5",
    "@types/node": "20.1.7",
    "@types/react": "18.2.6",
    "@types/react-dom": "18.2.4",
    "autoprefixer": "10.4.14",
    "class-variance-authority": "^0.6.1",
    "clsx": "^1.2.1",
    "contentlayer": "^0.3.2",
    "eslint": "8.40.0",
    "eslint-config-next": "13.4.2",
    "gray-matter": "^4.0.3",
    "lucide-react": "^0.252.0",
    "mdast-util-gfm": "^3.0.0",
    "mdast-util-gfm-table": "^2.0.0",
    "mdast-util-mdx": "^3.0.0",
    "next": "^13.4.19",
    "postcss": "8.4.23",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-redux": "^8.0.5",
    "rehype-autolink-headings": "^6.1.1",
    "rehype-pretty-code": "^0.9.5",
    "rehype-slug": "^5.1.0",
    "remark": "^14.0.3",
    "remark-code-frontmatter": "^1.0.0",
    "remark-gfm": "^3.0.1",
    "remark-html": "^15.0.2",
    "remark-mdx": "^2.3.0",
    "remark-rehype": "^10.1.0",
    "remark-slate-transformer": "^0.7.4",
    "shiki": "^0.14.2",
    "tailwind-merge": "^1.13.2",
    "tailwindcss": "3.3.2",
    "tailwindcss-animate": "^1.0.6",
    "typescript": "5.0.4"
  },
  "devDependencies": {
    "@tailwindcss/typography": "^0.5.9",
    "remark-cli": "^11.0.0",
    "remark-preset-lint-consistent": "^5.1.2",
    "remark-preset-lint-recommended": "^6.1.3",
    "remark-toc": "^8.0.1"
  }
}
``` 

Filecontents:

```
# GFM

## Autolink literals

www.example.com, https://example.com, and contact@example.com.

## Footnote

A note[^1]

[^1]: Big note.

## Strikethrough

~one~ or ~~two~~ tildes.

## Table

| a   | b   |   c |  d  |
| --- | :-- | --: | :-: |

## Tasklist

- [ ] to do
- [x] done
```

**Reply by ChristianMurphy** [reply]
@nicolasjorgevera it sounds like you are trying to do a mix of serverside and clientside rendering with next.js?
Could you share a runnable example of what you're trying? Even with the dependencies, there are a lot of pieces of your setup which are missing which make it hard to replicate.
https://next.new is a good sandbox started focus on reducing it to an MRE https://stackoverflow.com/help/minimal-reproducible-example

**Reply by nicolasjorgevera** [reply]
Hi Christian, basicly I try to mount a blog editor. That function is call to read an especific post in a .md file, transform the markdown to mdast tree (JSON) and return it. All this actions occurs in the server side. It's work fine without GFM extentions.

Unafortunely, this not work on stackblitz (I tryed) but I deployed it on Vercel and works (always without GFM).

If you like to see all the code and this "demo" on vercel you let me know.

**Reply by ChristianMurphy** [reply]
> Unafortunely, this not work on stackblitz (I tryed) but I deployed it on Vercel and works (always without GFM).

If the issue can't be replicated outside of Vercel/Next, it sounds like a bug in the Next framework/build tool itself, it works fine with other frameworks.
The Next community may be able to offer better support https://github.com/vercel/next.js/discussions
Though without a MRE, it may be challenging for them as well.

> transform the markdown to mdast tree (JSON) and return it. 

Using the AST as a wire protocol probably isn't what you are looking for, see the discussion in https://github.com/orgs/unifiedjs/discussions/188#discussioncomment-2466114
Consider rendering fully server side or fully client side, rather than a bespoke mix.

**Reply by wooorm** [reply, marked as answer]
You’re probably getting this error because some dependencies are out of date. Make sure every micromark and mdast package is at the latest.

**Reply by nicolasjorgevera** [reply]
@wooorm I update de depencencies and work! Thanks!!!

Thank you all!!
